rpcs3/rpcs3qt/syntax_highlighter.cpp

```diff
@@ -0,0 +1,129 @@
+
+void syntax_highlighter::highlightBlock(const QString &text)
+{
+	m_current_block = text;
+
+	for (const auto& rule : m_rules)
+	{
+		for (const auto& expression : rule.expressions)
+		{
+			// Search for all the matching strings
+			QRegularExpressionMatchIterator iter = expression.globalMatch(m_current_block);
+
+			// Iterate through the matching strings
+			while (iter.hasNext())
+			{
+				// Apply format to the matching string
+				QRegularExpressionMatch match = iter.next();
+				setFormat(match.capturedStart(), match.capturedLength(), rule.format);
+			}
+		}
+	}
+
+	for (const auto& rule : m_multi_rules)
+	{
+		// Search for all the matching strings
+		QRegularExpressionMatchIterator iter = rule.expression.globalMatch(m_current_block);
+
+		// Iterate through the matching strings
+		while (iter.hasNext())
+		{
+			// Apply formats to their respective found groups
+			QRegularExpressionMatch match = iter.next();
+			for (int i = 0; i <= match.lastCapturedIndex(); i++)
+			{
+				setFormat(match.capturedStart(i), match.capturedLength(i), rule.formats[i]);
+			}
+		}
+	}
+
+	for (const auto& rule : m_comment_rules)
+	{
+		int comment_start  = 0; // Current comment's start position in the text block
+		int comment_end    = 0; // Current comment's end position in the text block
+		int comment_length = 0; // Current comment length
+
+		// We assume we end outside a comment until we know better
+		setCurrentBlockState(ENDED_OUTSIDE_COMMENT);
+
+		// Search for the first comment in this block if we start outside or don't want to search for multiline comments
+		if (!rule.multi_line || previousBlockState() != ENDED_INSIDE_COMMENT)
+		{
+			comment_start = m_current_block.indexOf(rule.start_expression);
+		}
+
+		// Set format for the rest of this block/line
+		if (!rule.multi_line)
+		{
+			comment_length = m_current_block.length() - comment_start;
+			setFormat(comment_start, comment_length, rule.format);
+			break;
+		}
+
+		// Format all comments in this block (if they exist)
+		while (comment_start >= 0)
+		{
+			// Search for end of comment in the remaining text
+			QRegularExpressionMatch match = rule.end_expression.match(m_current_block, comment_start);
+			comment_end = match.capturedStart();
+			match.captured(1);
+
+			if (comment_end == -1)
+			{
+				// We end inside a comment and want to format the entire remaining text
+				setCurrentBlockState(ENDED_INSIDE_COMMENT);
+				comment_length = m_current_block.length() - comment_start;
+			}
+			else
+			{
+				// We found the end of the comment so we need to go another round
+				comment_length = comment_end - comment_start + match.capturedLength();
+			}
+
+			// Set format for this text segment
+			setFormat(comment_start, comment_length, rule.format);
+
+			// Search for the next comment
+			comment_start = m_current_block.indexOf(rule.start_expression, comment_start + comment_length);
+		}
+	}
+}
+
+bool syntax_highlighter::ExpressionInvalid(const QRegularExpression& expression)
```

Comment:
I'd name it `isInvalidExpression`, so it reads better, but otherwise looks fine.